plugins/source/pagerduty/client/client.go

```diff
@@ -0,0 +1,103 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/PagerDuty/go-pagerduty"
+	"github.com/cloudquery/plugin-sdk/schema"
+	"github.com/cloudquery/plugin-sdk/specs"
+	"github.com/rs/zerolog"
+	"gopkg.in/yaml.v3"
+)
+
+// By default, the maximum `limit` parameter allowed by pagerduty API is 100.
+const MaxPaginationLimit = 100
+
+type Client struct {
+	PagerdutyClient *pagerduty.Client
+	logger          zerolog.Logger
+}
+
+func Configure(ctx context.Context, logger zerolog.Logger, spec specs.Source) (schema.ClientMeta, error) {
+
+	authToken, err := getAuthToken()
+	if err != nil {
+		return nil, err
+	}
+
+	pagerdutyClient := pagerduty.NewClient(authToken)
+
+	cqClient := Client{
+		PagerdutyClient: pagerdutyClient,
+		logger:          logger,
+	}
+
+	return &cqClient, nil
+}
+
+func (c *Client) Logger() *zerolog.Logger {
+	return &c.logger
+}
+
+type PdYmlStruct struct {
+	Authtoken string
+}
+
+// There is no easy way to get account-id or similar from the pagerduty API.
+func (c *Client) ID() string {
+	return "pagerduty"
+}
+
+// getAuthToken returns the pagerduty auth token.
+// Is supports the following methods (in the following order of precedence):
+// - Reading from `PAGERDUTY_AUTH_TOKEN` environment variable.
+// - Reading from `~/.pd.yml` or `~/.pd.yaml` file. (Similar to the pagerduty CLI).
+func getAuthToken() (string, error) {
+	failedToGetAuthTokenError := fmt.Errorf("failed to get pagerduty auth token. Please provide an auth-token (see https://www.cloudquery.io/docs/plugins/sources/pagerduty/overview#authentication)")
+
+	envAuthToken := os.Getenv("PAGERDUTY_AUTH_TOKEN")
+	if envAuthToken != "" {
+		return envAuthToken, nil
+	}
+
+	yamlFileContents := getPdYamlFileContents()
+	if len(yamlFileContents) == 0 {
+		return "", failedToGetAuthTokenError
+	}
+
+	var pdYmlStruct PdYmlStruct
+	err := yaml.Unmarshal(yamlFileContents, &pdYmlStruct)
+	if err != nil {
+		return "", fmt.Errorf("failed to unmarshal `.pd.yml` file: %w", err)
+	}
+
+	if len(pdYmlStruct.Authtoken) == 0 {
+		return "", failedToGetAuthTokenError
+	}
+
+	return pdYmlStruct.Authtoken, nil
+}
+
+// gets the contents of `~/.pd.yml` or `~/.pd.yaml`.
+// Otherwise - returns nil.
+func getPdYamlFileContents() []byte {
+	home, err := os.UserHomeDir()
+	if err != nil {
+		return nil
+	}
+
+	contents, error := os.ReadFile(filepath.Join(home, ".pd.yml"))
```

Comment:
Not a big deal necessarily, but I think this function could hide errors where the file exists, but the user doesn't have access to it, and maybe some other types of non-obvious errors. Since this is our last attempt in the order of presedence, maybe we can return the error related to opening `.pd.yml` (or whatever the preferred name is) so that the user can act on it? This could be in addition to the error message defined higher up https://github.com/cloudquery/cloudquery/pull/5547/files#diff-576bc1129ebe50277aa54dcd2c7863d410c1c4c6415527263b9824b9c0e80249R59